gob fairness

Hi.
Why not using the same instance of gob Encoder/Decoder, instead of creating a new one for each marshall/unmarshall operation? Seems to me that it would be more fair... what protobuf does once at generate time, gob can do at runtime, but just once! Otherwise consider removing gob from the benchmark, it makes it seem really bad while it probably isn't.

I'd welcome a PR for this. I [tried](https://github.com/alecthomas/go_serialization_benchmarks/tree/faster_gob/36), but kept getting corrupt message errors:

```
BenchmarkGobMarshal-8  	 1000000	      1207 ns/op	      48 B/op	       2 allocs/op
BenchmarkGobUnmarshal-8	b=61
--- FAIL: BenchmarkGobUnmarshal-8
	serialization_benchmarks_test.go:114: gob failed to unmarshal: gob: unknown type id or corrupted data (<��b9b2b4faa77947cf�4���IL�
		1ca61f574b��Y�K���?)
ok  	github.com/alecthomas/go_serialization_benchmarks	1.244s
```

--- a/serialization_benchmarks_test.go
+++ b/serialization_benchmarks_test.go
@@ -10,6 +10,8 @@ import (
 	"testing"
 	"time"
 
+	"zombiezen.com/go/capnproto2"
+
 	"github.com/DeDiS/protobuf"
 	"github.com/Sereal/Sereal/Go/sereal"
 	"github.com/alecthomas/binary"
@@ -17,13 +19,12 @@ import (
 	"github.com/glycerine/go-capnproto"
 	"github.com/gogo/protobuf/proto"
 	flatbuffers "github.com/google/flatbuffers/go"
-	"github.com/hprose/hprose-go/io"
+	"github.com/hprose/hprose-go"
 	"github.com/tinylib/msgp/msgp"
 	"github.com/ugorji/go/codec"
 	vitessbson "github.com/youtube/vitess/go/bson"
 	"gopkg.in/mgo.v2/bson"
 	vmihailenco "gopkg.in/vmihailenco/msgpack.v2"
-	"zombiezen.com/go/capnproto2"
 )
 
 var (
@@ -98,7 +99,10 @@ func benchUnmarshal(b *testing.B, s Serializer) {
 	data := generate()
 	ser := make([][]byte, len(data))
 	for i, d := range data {
-		ser[i] = []byte(string(s.Marshal(d)))
+		o := s.Marshal(d)
+		t := make([]byte, len(o))
+		copy(t, o)
+		ser[i] = t
 	}
 	b.ReportAllocs()
 	b.StartTimer()
@@ -255,22 +259,26 @@ func BenchmarkVitessBsonUnmarshal(b *testing.B) {
 
 // encoding/gob
 
-type GobSerializer struct{}
+type GobSerializer struct {
+	b   *bytes.Buffer
+	enc *gob.Encoder
+	dec *gob.Decoder
+}
 
-func (g GobSerializer) Marshal(o interface{}) []byte {
-	b := &bytes.Buffer{}
-	e := gob.NewEncoder(b)
-	err := e.Encode(o)
+func (g *GobSerializer) Marshal(o interface{}) []byte {
+	g.b.Reset()
+	err := g.enc.Encode(o)
 	if err != nil {
 		panic(err)
 	}
-	return b.Bytes()
+	return g.b.Bytes()
 }
 
-func (g GobSerializer) Unmarshal(d []byte, o interface{}) error {
-	b := bytes.NewBuffer(d)
-	e := gob.NewDecoder(b)
-	err := e.Decode(o)
+func (g *GobSerializer) Unmarshal(d []byte, o interface{}) error {
+	g.b.Reset()
+	g.b.Write(d)
+	fmt.Printf("b=%d\n", g.b.Len())
+	err := g.dec.Decode(o)
 	return err
 }
 
@@ -279,11 +287,21 @@ func (g GobSerializer) String() string {
 }
 
 func BenchmarkGobMarshal(b *testing.B) {
-	benchMarshal(b, GobSerializer{})
+	buf := &bytes.Buffer{}
+	benchMarshal(b, &GobSerializer{
+		b:   buf,
+		enc: gob.NewEncoder(buf),
+		dec: gob.NewDecoder(buf),
+	})
 }
 
 func BenchmarkGobUnmarshal(b *testing.B) {
-	benchUnmarshal(b, GobSerializer{})
+	buf := &bytes.Buffer{}
+	benchUnmarshal(b, &GobSerializer{
+		b:   buf,
+		enc: gob.NewEncoder(buf),
+		dec: gob.NewDecoder(buf),
+	})
 }
 
 // github.com/davecgh/go-xdr/xdr

--- a/serialization_benchmarks_test.go
+++ b/serialization_benchmarks_test.go
@@ -289,6 +289,8 @@ func BenchmarkGobMarshal(b *testing.B) {
 	var s GobSerializer
 	s.enc = gob.NewEncoder(&s.b)
 	s.dec = gob.NewDecoder(&s.b)
+	// Decode and encode A once,
+	// so that type information is transmitted correctly.
 	err := s.enc.Encode(A{})
 	if err != nil {
 		panic(err)

--- a/serialization_benchmarks_test.go
+++ b/serialization_benchmarks_test.go
@@ -260,7 +260,7 @@ func BenchmarkVitessBsonUnmarshal(b *testing.B) {
 // encoding/gob
 
 type GobSerializer struct {
-	b   *bytes.Buffer
+	b   bytes.Buffer
 	enc *gob.Encoder
 	dec *gob.Decoder
 }
@@ -277,7 +277,6 @@ func (g *GobSerializer) Marshal(o interface{}) []byte {
 func (g *GobSerializer) Unmarshal(d []byte, o interface{}) error {
 	g.b.Reset()
 	g.b.Write(d)
-	fmt.Printf("b=%d\n", g.b.Len())
 	err := g.dec.Decode(o)
 	return err
 }
@@ -287,21 +286,35 @@ func (g GobSerializer) String() string {
 }
 
 func BenchmarkGobMarshal(b *testing.B) {
-	buf := &bytes.Buffer{}
-	benchMarshal(b, &GobSerializer{
-		b:   buf,
-		enc: gob.NewEncoder(buf),
-		dec: gob.NewDecoder(buf),
-	})
+	var s GobSerializer
+	s.enc = gob.NewEncoder(&s.b)
+	s.dec = gob.NewDecoder(&s.b)
+	err := s.enc.Encode(A{})
+	if err != nil {
+		panic(err)
+	}
+	var a A
+	err = s.dec.Decode(&a)
+	if err != nil {
+		panic(err)
+	}
+	benchMarshal(b, &s)
 }
 
 func BenchmarkGobUnmarshal(b *testing.B) {
-	buf := &bytes.Buffer{}
-	benchUnmarshal(b, &GobSerializer{
-		b:   buf,
-		enc: gob.NewEncoder(buf),
-		dec: gob.NewDecoder(buf),
-	})
+	var s GobSerializer
+	s.enc = gob.NewEncoder(&s.b)
+	s.dec = gob.NewDecoder(&s.b)
+	err := s.enc.Encode(A{})
+	if err != nil {
+		panic(err)
+	}
+	var a A
+	err = s.dec.Decode(&a)
+	if err != nil {
+		panic(err)
+	}
+	benchUnmarshal(b, &s)
 }
 
 // github.com/davecgh/go-xdr/xdr

